src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -2145,13 +2175,28 @@ internal static object GetExpressionValue(ExpressionAst expressionAst, bool isTr
                 var pipe = new Pipe(resultList);
                 try
                 {
+                    ScriptBlock scriptBlock = null;
+                    if (ProfilerEventSource.LogInstance.IsEnabled())
+                    {
+                        // We need a scriptblock Id to keep track of the call stack.
+                        scriptBlock = ScriptBlock.Create(context, expressionAst.Extent.Text);
+
+                        // No need to optimize - we only want to add to the scriptblock cache
+                        // so that we can get rundown events.
+                        // We can safely do this because lambda and sequencePoints
+                        // parameters of the method come always from callsites as null.
+                        // (We could remove them at all.)
+                        scriptBlock.Compile(optimized: false);
```

Comment:
Is this for getting the script call stack when generating events? It looks quite hacky.

I think we need to define the goal for this PR. I guess the goal is to make the Pester's coverage functionality able to work with the ETW events generated from the changes here. So, is the script stack necessary for the Pester's coverage functionality to work?

If we set that as the goal, we should only include features/changes that are necessary for that to work.

Reply:
Both Pester code coverage, and Profiler would depend on this. Pester code coverage is a subset of Profiler functionality. Both use this interface to provide their services: 

https://github.com/nohwnd/Profiler/blob/main/csharp/Profiler/ITracer.cs

Which collects: 

// The currently executing scriptblock and extent
IScriptExtent extent, 
ScriptBlock scriptBlock, 

// how many items are on callstack
int level, 

// the current function and module name
string functionName, 
string moduleName

The data is collected here, by reflecting on debugger, functionContext and session state: 

https://github.com/nohwnd/Profiler/blob/main/csharp/Profiler/Tracer.cs#L157-L171

As long as this info is provided in some form, the current version of Profiler can work with this. 

"Level" I should be able to get from scriptblock ids as pointed out here: https://github.com/PowerShell/PowerShell/pull/13673#issuecomment-831161136

Module I should be able to get from the scriptblock and its associated session state. I think.

But what about the function name? Any suggestions where I can grab that?